<a href="https://colab.research.google.com/github/herlocksholmes1888/CienciaDeDados/blob/main/Curso_B%C3%A1sico_de_Ci%C3%AAncia_de_Dados_Aula_04.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [34]:
import pandas as pd
import io

In [2]:
from google.colab import files
uploaded = files.upload()

Saving titanic_train.csv to titanic_train.csv


In [3]:
titanic = pd.read_csv(io.BytesIO(uploaded['titanic_train.csv']))
titanic

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

In [4]:
# axis=1 elimina uma coluna, axis=0 elimina uma linha
titanic = titanic.drop('Cabin', axis=1)

In [5]:
titanic.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(4)
memory usage: 76.7+ KB


In [8]:
titanic = titanic.dropna()

In [9]:
titanic.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 712 entries, 0 to 890
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  712 non-null    int64  
 1   Survived     712 non-null    int64  
 2   Pclass       712 non-null    int64  
 3   Name         712 non-null    object 
 4   Sex          712 non-null    object 
 5   Age          712 non-null    float64
 6   SibSp        712 non-null    int64  
 7   Parch        712 non-null    int64  
 8   Ticket       712 non-null    object 
 9   Fare         712 non-null    float64
 10  Embarked     712 non-null    object 
dtypes: float64(2), int64(5), object(4)
memory usage: 66.8+ KB


In [10]:
import numpy as np

In [20]:
titanic['Sex'] = np.where(titanic['Sex']=="male", 1, 0)

In [13]:
titanic

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
885          886         0       3   
886          887         0       2   
887          888         1       1   
889          890         1       1   
890          891         0       3   

                                                  Name  Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    1  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...    0  38.0      1   
2                               Heikkinen, Miss. Laina    0  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)    0  35.0      1   
4                             Allen, Mr. William Henry    1  35.0      0   
..                                                 ...  ...   ...    ...   
885               Rice, Mrs. William (Margaret Norton)    0  39.0      0   
886                              Montvila, Rev. Juozas    1  27.0      0   
887                       Graham, Miss. Margaret Edith    0  19.0      0   
889                              Behr, Mr. Karl Howell    1  26.0      0   
890                                Dooley, Mr. Patrick    1  32.0      0   

     Parch            Ticket     Fare Embarked  
0        0         A/5 21171   7.2500        S  
1        0          PC 17599  71.2833        C  
2        0  STON/O2. 3101282   7.9250        S  
3        0            113803  53.1000        S  
4        0            373450   8.0500        S  
..     ...               ...      ...      ...  
885      5            382652  29.1250        Q  
886      0            211536  13.0000        S  
887      0            112053  30.0000        S  
889      0            111369  30.0000        C  
890      0            370376   7.7500        Q  

[712 rows x 11 columns]

In [17]:
columns = titanic.dtypes[titanic.dtypes.values == 'object']
columns

Series([], dtype: object)

In [19]:
titanic = titanic.drop(columns, axis=1)
titanic.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 712 entries, 0 to 890
Data columns (total 8 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  712 non-null    int64  
 1   Survived     712 non-null    int64  
 2   Pclass       712 non-null    int64  
 3   Sex          712 non-null    int64  
 4   Age          712 non-null    float64
 5   SibSp        712 non-null    int64  
 6   Parch        712 non-null    int64  
 7   Fare         712 non-null    float64
dtypes: float64(2), int64(6)
memory usage: 50.1 KB


In [21]:
X = titanic.drop('Survived', axis=1)
Y = titanic.Survived

Treinamento de modelos diversos para previsão do índice de sobrevivência.

In [22]:
from sklearn.neighbors import KNeighborsClassifier
neigh = KNeighborsClassifier(n_neighbors=3)

In [23]:
neigh.fit(X, Y)

KNeighborsClassifier(n_neighbors=3)

In [24]:
neigh.score(X, Y)

0.7907303370786517

In [76]:
from sklearn import tree

In [77]:
clfTree = tree.DecisionTreeClassifier()

In [78]:
clfTree = clfTree.fit(X, Y)

In [79]:
clfTree.score(X, Y)

1.0

Um problema comum na árvore de decisão é que ela tem o hábito de decorar os dados ao invés de aprendê-los. Para verificar se foi esse o caso, é bom se utilizar de dados de teste. 

In [30]:
from sklearn.linear_model import LogisticRegression

In [72]:
clf = LogisticRegression(random_state=0, max_iter=1000).fit(X, Y)

In [73]:
clf.score(X, Y)

0.7064606741573034

Fase de testes

In [59]:
from google.colab import files
uploaded = files.upload()

Saving titanic_test.csv to titanic_test.csv


In [60]:
test = pd.read_csv(io.BytesIO(uploaded['titanic_test.csv']))

In [61]:
test = test.drop('Cabin', axis=1)
test = test.dropna()
test['Sex'] = np.where(test['Sex']=="male", 1, 0)
columns = test.dtypes[test.dtypes.values == 'object'].index

In [62]:
test = test.drop(columns, axis=1)

In [63]:
test.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 331 entries, 0 to 415
Data columns (total 8 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  331 non-null    int64  
 1   Survived     331 non-null    int64  
 2   Pclass       331 non-null    int64  
 3   Sex          331 non-null    int64  
 4   Age          331 non-null    float64
 5   SibSp        331 non-null    int64  
 6   Parch        331 non-null    int64  
 7   Fare         331 non-null    float64
dtypes: float64(2), int64(6)
memory usage: 23.3 KB


In [67]:
X_test = test.drop('Survived', axis=1)
Y_test = test.Survived

In [80]:
pred_KNN = neigh.predict(X_test)
tree = clfTree.predict(X_test)
log = clf.predict(X_test)

Matriz de confusão

In [66]:
from sklearn.metrics import confusion_matrix

In [69]:
confusion_matrix(Y_test, pred_KNN)

array([[ 68, 136],
       [ 35,  92]])

Os valores tidos como 0 e que foram vistos como 0 foram 68, os valores tidos como 0 e que foram vistos como 1 foram 136. Os valores tidos como 1 e que foram vistos como 0 foram 35, e os valores tidos como 1 e que foram vistos como 1 foram 92.

In [81]:
confusion_matrix(Y_test, tree)

array([[111,  93],
       [ 43,  84]])

In [82]:
confusion_matrix(Y_test, log)

array([[131,  73],
       [ 64,  63]])

**Acurácia**

In [83]:
from sklearn.metrics import accuracy_score

In [84]:
accuracy_score(Y_test, pred_KNN)

0.48338368580060426

In [85]:
accuracy_score(Y_test, tree)

0.5891238670694864

In [86]:
accuracy_score(Y_test, log)

0.5861027190332326

**Precisão**

In [87]:
from sklearn.metrics import precision_score

In [88]:
precision_score(Y_test, pred_KNN)

0.40350877192982454

In [89]:
precision_score(Y_test, tree)

0.4745762711864407

In [90]:
precision_score(Y_test, log)

0.4632352941176471

In [91]:
from sklearn.metrics import recall_score

In [92]:
recall_score(Y_test, pred_KNN)

0.7244094488188977

In [93]:
recall_score(Y_test, tree)

0.6614173228346457

In [94]:
recall_score(Y_test, log)

0.49606299212598426

Qual das métricas será escolhida dependerá do que o cientista/analista deseja. Ele gostaria de saber quais pessoas sobreviveram precisamente? É melhor usar recall. Você quer ponderar quantas pessoas morreram e quantas sobreviveram? Talvez seja melhor reconsiderar.